cql3/query_processor.cc

```diff
@@ -1025,16 +1026,30 @@ query_processor::execute_schema_statement(const statements::schema_altering_stat
 
     cql3::cql_warnings_vec warnings;
 
+    auto request_id = guard->new_group0_state_id();
+    // TODO: either save as a member of statements::schema_altering_statement class
+    //       or extend alter_keyspace_statement::prepare_schema_mutations with guard's req id
+    const_cast<statements::schema_altering_statement&>(stmt).global_req_id = fmt::to_string(request_id);
```

Comment:
why is it converted to a string and not kept as uuid?

(orthogonal to this series why is the new group state id a generic UUID and not a strongly typed one? Cc @kbr-scylla )

Reply:
As far as I recall, because I couldn't easily (de)serialize uuid.

Reply:
UUIDs should (de)serialize without problems. If you encountered a problem you can ask for help and show us the failure (compilation failure?)

Also, `const_cast` is a footgun. Did you make sure it is safe to use in this context? Can it be avoided?

Reply:
I managed to change `global_req_id` to be of `utils::UUID` type.

> Also, `const_cast` is a footgun. Did you make sure it is safe to use in this context? Can it be avoided?

@kbr-scylla the reason for a const_cast is that `stmt` is `const`, I'd need to change the signature of this function (it takes `const stmt` as an arg) to skip `const_cast`, but that'd result in a chain of interface changes, and because I didn't want to introduce too many changes I decided to go with the const_cast. But I'm not sure what's be best move here, hence the TODO comment. I'm opened to suggestions. If there're none, I'll remove the comment and leave the const_cast.

Reply:
I removed attached TODO comment and left the const_cast. If you have other idea, let me know. As of now I'm resolving the conversation.

Reply:
I googled about `const_cast`, and people on the internet claim that what you're doing here is **undefined behavior**. Unfortunately I don't have time and energy to look into the standard and verify their claims.

For example
https://stackoverflow.com/questions/357600/is-const-cast-safe

one of the commenter posted this paragraph (dunno which version of the standard it comes from so maybe it's outdated -- but it should be verified)

> §7.1.​5.1/4 says Except that any class member declared mutable (7.1.1) can be modified, any attempt to modify a const object during its lifetime (3.8) results in undefined behavior

Reply:
> 
> > §7.1.​5.1/4 says Except that any class member declared mutable (7.1.1) can be modified, any attempt to modify a const object during its lifetime (3.8) results in undefined behavior

Well, C++ is a low level language. A const object can be in ROM.

Reply:
If I understand correctly you're basically using the statement object to pass a parameter to a function you call further down the chain.

If you don't want to extend the function with additional parameters (I guess it's a large interface change?) then maybe a workaround which should not introduce UB would be to mark `global_req_id` as `mutable` (although that is not the purpose of `mutable`, or `const_cast`, but meh)

Reply:
https://en.cppreference.com/w/cpp/language/const_cast says:
```
Modifying a const object through a non-const access path and referring to a volatile object through a non-volatile [glvalue](https://en.cppreference.com/w/cpp/language/value_category#glvalue) results in undefined behavior.
```

Reply:
> If I understand correctly you're basically using the statement object to pass a parameter to a function you call further down the chain.
> 
> If you don't want to extend the function with additional parameters (I guess it's a large interface change?) then maybe a workaround which should not introduce UB would be to mark `global_req_id` as `mutable` (although that is not the purpose of `mutable`, or `const_cast`, but meh)

Alternatively, you can implement a `set_global_request_id` method that would specified as const, but would set the _global_req_id member, that can be defined as mutable for this purpose.

Reply:
I changed `global_req_id` to be `mutable`.
I don't think a method is needed, as this member is both set and read globally, so get() + set() would only serve as wrappers for a `public` data member.
I was aware about UB when `const_cast`ing, but I've never experienced any problems with it, our code is full of const_casts and it was only a temporary solution marked with a TODO comment originally.
Either way, I'll make sure to refactor this in a follow up PR. Thanks for all the suggestions. For now I think we're good, because the UB has been removed.